File: app/frontend/src/contractFields.js
/** Sea contract tab — saved JSON on candidate (contract_json). */

/** Candidate profile fields available in docxtpl (from CRM card, not contract_json). */
export const CANDIDATE_PERSONAL_PLACEHOLDER_DEFS = [
  { key: "current_date", label: "Current date" },
  { key: "surname", label: "Фамилия" },
  { key: "first_name", label: "Имя" },
  { key: "middle_name", label: "Отчество" },
  { key: "full_name", label: "Полное имя" },
  { key: "latin_full_name", label: "Latin Full Name" },
  { key: "native_full_name", label: "Native Full Name" },
  { key: "date_of_birth", label: "Дата рождения" },
  { key: "place_of_birth", label: "Место рождения" },
  { key: "country_of_birth", label: "Страна рождения" },
  { key: "nationality", label: "Национальность" },
  { key: "citizenship", label: "Гражданство" },
  { key: "age", label: "Возраст" },
  { key: "gender", label: "Пол" },
  { key: "marital_status", label: "Семейное положение" },
  { key: "father_name", label: "Имя отца" },
  { key: "mother_name", label: "Имя матери" },
  { key: "primary_phone", label: "Основной телефон" },
  { key: "mobile_phone", label: "Мобильный" },
  { key: "email", label: "Email" },
  { key: "home_address", label: "Домашний адрес" },
  { key: "permanent_address", label: "Постоянный адрес" },
  { key: "city", label: "Город" },
  { key: "country", label: "Страна" },
  { key: "current_rank", label: "Текущая должность" },
  { key: "passport_number", label: "Номер паспорта" },
  { key: "passport_issue_date", label: "Дата выдачи паспорта" },
  { key: "passport_expiry_date", label: "Срок действия паспорта" },
  { key: "passport_place_of_issue", label: "Кем / где выдан паспорт" },
  { key: "seaman_book_number", label: "Seaman's Book — номер" },
  { key: "seaman_book_issue_date", label: "Seaman's Book — дата выдачи" },
  { key: "seaman_book_expiry_date", label: "Seaman's Book — срок действия" },
];

export const CANDIDATE_PERSONAL_PLACEHOLDER_LINES = CANDIDATE_PERSONAL_PLACEHOLDER_DEFS.map(
  ({ key }) => `{{ ${key} }}`
);

export const CONTRACT_EDITABLE_FIELD_DEFS = [
  { key: "contract_sign_date", label: "Дата подписания контракта", placeholder: "contract_sign_date", type: "date" },
  { key: "contract_period", label: "Срок контракта / Period of employment", placeholder: "contract_period", type: "text" },
  { key: "contract_embarkation_date", label: "Дата посадки", placeholder: "contract_embarkation_date", type: "date" },
  { key: "contract_embarkation_port", label: "Порт посадки", placeholder: "contract_embarkation_port", type: "text" },
  { key: "contract_number", label: "Номер контракта", placeholder: "contract_number", type: "text" },
  { key: "contract_remarks", label: "Примечания", placeholder: "contract_remarks", type: "text" },
];

export const CONTRACT_DEPARTURE_FIELD_DEFS = [
  {
    key: "contract_departure_date",
    label: "Дата вылета",
    placeholder: "contract_departure_date",
    type: "date",
  },
];

export const CONTRACT_AIRPORT_FIELD_DEFS = [
  { key: "contract_home_airport", label: "Home airport", placeholder: "contract_home_airport", type: "text" },
  {
    key: "contract_departure_airport",
    label: "Departure airport",
    placeholder: "contract_departure_airport",
    type: "text",
  },
];

export const CONTRACT_ALL_EDITABLE_FIELD_DEFS = [
  ...CONTRACT_EDITABLE_FIELD_DEFS,
  ...CONTRACT_DEPARTURE_FIELD_DEFS,
  ...CONTRACT_AIRPORT_FIELD_DEFS,
];

export const CONTRACT_VESSEL_PREVIEW_FIELDS = [
  ["name", "Название"],
  ["imo", "IMO"],
  ["flag", "Флаг"],
  ["port_of_registry", "Port of Registry"],
  ["grt", "GRT"],
  ["deadweight", "Dead Weight"],
  ["year_built", "Year of Built"],
  ["engine_type", "Engine Type"],
  ["engine_hp", "H.P."],
];

export const CONTRACT_PLACEHOLDER_LINES = [
  "{{ contract_company_name }}",
  "{{ contract_company_slug }}",
  "{{ contract_vessel_name }}",
  "{{ contract_rank }}",
  ...CONTRACT_ALL_EDITABLE_FIELD_DEFS.map((f) => `{{ ${f.placeholder} }}`),
  "{{ home_airport }}",
  "{{ departure_airport }}",
  "{{ departure_date }}",
  "{{ contract_vessel_imo }}",
  "{{ contract_vessel_flag }}",
  "{{ contract_vessel_grt }}",
  "{{ contract_vessel_type }}",
  "{{ salary_total_wage }}",
  "{{ salary_basic_monthly_wage }}",
];

export function createEmptyContractForm() {
  return {
    company_id: "",
    vessel_id: "",
    rank: "",
    ...Object.fromEntries(CONTRACT_ALL_EDITABLE_FIELD_DEFS.map(({ key }) => [key, ""])),
  };
}

export function hydrateContractForm(parsed, candidate) {
  const next = { ...parsed };
  for (const { key } of CONTRACT_AIRPORT_FIELD_DEFS) {
    if (String(next[key] ?? "").trim()) {
      continue;
    }
    if (key === "contract_home_airport" && candidate?.home_airport) {
      next[key] = String(candidate.home_airport);
    }
    if (key === "contract_departure_airport" && candidate?.departure_airport) {
      next[key] = String(candidate.departure_airport);
    }
  }
  return next;
}

export function parseContractJson(raw) {
  const empty = createEmptyContractForm();
  if (!raw || typeof raw !== "string" || !raw.trim()) {
    return empty;
  }
  try {
    const data = JSON.parse(raw);
    if (!data || typeof data !== "object") {
      return empty;
    }
    const next = { ...empty };
    next.company_id = data.company_id != null ? String(data.company_id) : "";
    next.vessel_id = data.vessel_id != null ? String(data.vessel_id) : "";
    next.rank = data.rank || "";
    for (const { key } of CONTRACT_ALL_EDITABLE_FIELD_DEFS) {
      next[key] = data[key] == null ? "" : String(data[key]);
    }
    return next;
  } catch {
    return empty;
  }
}

export function contractFormToApiPayload(form) {
  const companyId = Number(form.company_id);
  const vesselRaw = String(form.vessel_id ?? "").trim();
  const payload = {
    company_id: companyId,
    vessel_id: vesselRaw === "" ? null : Number(vesselRaw),
    rank: String(form.rank || "").trim(),
  };
  for (const { key } of CONTRACT_ALL_EDITABLE_FIELD_DEFS) {
    const value = String(form[key] ?? "").trim();
    payload[key] = value === "" ? null : value;
  }
  return payload;
}

export function salaryMatchesContract(form, salaryJson) {
  if (!salaryJson || !form.company_id || !form.rank) {
    return false;
  }
  return (
    String(salaryJson.company_id) === String(form.company_id) &&
    String(salaryJson.rank || "").trim() === String(form.rank || "").trim()
  );
}

File: app/contract_fields.py
"""Sea contract tab: saved JSON on candidate and docxtpl placeholders."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from sqlalchemy.orm import Session

from app.vessel_specs import VESSEL_FIELD_SPECS, vessel_placeholder_suffix, vessel_placeholder_value
from models.schema import Company, Vessel

CONTRACTS_FOLDER_NAMES: frozenset[str] = frozenset({"контракты", "contracts"})

# Editable fields stored in contract_json (keys match placeholder names).
CONTRACT_EDITABLE_FIELDS: tuple[tuple[str, str], ...] = (
    ("contract_sign_date", "Дата подписания контракта"),
    ("contract_period", "Срок контракта / Period of employment"),
    ("contract_embarkation_date", "Дата посадки"),
    ("contract_embarkation_port", "Порт посадки"),
    ("contract_number", "Номер контракта"),
    ("contract_remarks", "Примечания"),
)

CONTRACT_DEPARTURE_FIELDS: tuple[tuple[str, str], ...] = (
    ("contract_departure_date", "Дата вылета"),
)

CONTRACT_AIRPORT_FIELDS: tuple[tuple[str, str], ...] = (
    ("contract_home_airport", "Home airport"),
    ("contract_departure_airport", "Departure airport"),
)

CONTRACT_ALL_EDITABLE_FIELDS: tuple[tuple[str, str], ...] = (
    *CONTRACT_EDITABLE_FIELDS,
    *CONTRACT_DEPARTURE_FIELDS,
    *CONTRACT_AIRPORT_FIELDS,
)

CONTRACT_SELECTION_KEYS = ("company_id", "company_name", "vessel_id", "vessel_name", "rank")

# Candidate profile fields in docxtpl context (CRM card, not contract_json).
CANDIDATE_PERSONAL_PLACEHOLDER_FIELDS: tuple[tuple[str, str], ...] = (
    ("current_date", "Current date"),
    ("surname", "Фамилия"),
    ("first_name", "Имя"),
    ("middle_name", "Отчество"),
    ("full_name", "Полное имя"),
    ("latin_full_name", "Latin Full Name"),
    ("native_full_name", "Native Full Name"),
    ("date_of_birth", "Дата рождения"),
    ("place_of_birth", "Место рождения"),
    ("country_of_birth", "Страна рождения"),
    ("nationality", "Национальность"),
    ("citizenship", "Гражданство"),
    ("age", "Возраст"),
    ("gender", "Пол"),
    ("marital_status", "Семейное положение"),
    ("father_name", "Имя отца"),
    ("mother_name", "Имя матери"),
    ("primary_phone", "Основной телефон"),
    ("mobile_phone", "Мобильный"),
    ("email", "Email"),
    ("home_address", "Домашний адрес"),
    ("permanent_address", "Постоянный адрес"),
    ("city", "Город"),
    ("country", "Страна"),
    ("current_rank", "Текущая должность"),
    ("passport_number", "Номер паспорта"),
    ("passport_issue_date", "Дата выдачи паспорта"),
    ("passport_expiry_date", "Срок действия паспорта"),
    ("passport_place_of_issue", "Кем / где выдан паспорт"),
    ("seaman_book_number", "Seaman's Book — номер"),
    ("seaman_book_issue_date", "Seaman's Book — дата выдачи"),
    ("seaman_book_expiry_date", "Seaman's Book — срок действия"),
)


def parse_contract_json(raw: str | None) -> dict[str, Any]:
    import json

    if not raw or not str(raw).strip():
        return {}
    try:
        data = json.loads(raw)
    except json.JSONDecodeError:
        return {}
    return data if isinstance(data, dict) else {}


def build_saved_contract_payload(
    *,
    company_id: int,
    company_name: str,
    vessel_id: int | None,
    vessel_name: str | None,
    rank: str,
    editable: dict[str, Any],
    username: str | None,
) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "company_id": company_id,
        "company_name": company_name,
        "vessel_id": vessel_id,
        "vessel_name": vessel_name or "",
        "rank": rank.strip(),
        "updated_at": datetime.utcnow().isoformat(timespec="seconds") + "Z",
        "updated_by": username or "",
    }
    for key, _label in CONTRACT_ALL_EDITABLE_FIELDS:
        value = editable.get(key)
        payload[key] = "" if value is None else str(value).strip()
    return payload


def _fmt(value: Any) -> str:
    if value is None:
        return ""
    return str(value).strip() if isinstance(value, str) else str(value)


def contract_placeholders_from_saved(
    data: dict[str, Any],
    db_session: Session | None = None,
) -> dict[str, str]:
    if not data:
        return {}

    placeholders: dict[str, str] = {
        "contract_company_name": _fmt(data.get("company_name")),
        "contract_vessel_name": _fmt(data.get("vessel_name")),
        "contract_rank": _fmt(data.get("rank")),
    }
    for key, _label in CONTRACT_ALL_EDITABLE_FIELDS:
        placeholders[key] = _fmt(data.get(key))

    home_airport = _fmt(data.get("contract_home_airport"))
    departure_airport = _fmt(data.get("contract_departure_airport"))
    departure_date = _fmt(data.get("contract_departure_date"))
    if home_airport:
        placeholders["home_airport"] = home_airport
    if departure_airport:
        placeholders["departure_airport"] = departure_airport
    if departure_date:
        placeholders["departure_date"] = departure_date

    company_id = data.get("company_id")
    vessel_id = data.get("vessel_id")
    if db_session is not None and company_id:
        company = db_session.get(Company, int(company_id))
        if company:
            placeholders["contract_company_name"] = company.name or ""
            placeholders["contract_company_slug"] = company.slug or ""

    if db_session is not None and vessel_id:
        vessel = db_session.get(Vessel, int(vessel_id))
        if vessel:
            placeholders["contract_vessel_name"] = vessel.name or placeholders["contract_vessel_name"]
            for field_key, _label in VESSEL_FIELD_SPECS:
                suffix = vessel_placeholder_suffix(field_key)
                placeholders[f"contract_vessel_{suffix}"] = vessel_placeholder_value(vessel, field_key)
            if company_id:
                company = db_session.get(Company, int(company_id))
                if company:
                    prefix = f"company_{company.slug}_{vessel.slug}"
                    for field_key, _label in VESSEL_FIELD_SPECS:
                        suffix = vessel_placeholder_suffix(field_key)
                        placeholders[f"{prefix}_{suffix}"] = vessel_placeholder_value(vessel, field_key)

    return placeholders


def candidate_personal_placeholder_lines() -> list[str]:
    return [f"{{{{ {key} }}}}" for key, _label in CANDIDATE_PERSONAL_PLACEHOLDER_FIELDS]


def contract_placeholder_lines() -> list[str]:
    lines = [
        "{{ contract_company_name }}",
        "{{ contract_company_slug }}",
        "{{ contract_vessel_name }}",
        "{{ contract_rank }}",
    ]
    for key, _label in CONTRACT_ALL_EDITABLE_FIELDS:
        lines.append(f"{{{{ {key} }}}}")
    lines.extend(["{{ home_airport }}", "{{ departure_airport }}", "{{ departure_date }}"])
    for field_key, _label in VESSEL_FIELD_SPECS:
        suffix = vessel_placeholder_suffix(field_key)
        lines.append(f"{{{{ contract_vessel_{suffix} }}}}")
    return lines
